jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,665 @@
+     * converted to a more efficient {@link #includeMethod(String...)} or {@link #include(PathSpec...)} respectively.
+     */
+    @SafeVarargs
+    public final void include(Predicate<Request>... predicates)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (Predicate<Request> p : predicates)
+        {
+            if (p instanceof MethodPredicate methodPredicate)
+                includeMethod(methodPredicate._method);
+            else if (p instanceof PathSpecPredicate pathSpecPredicate)
+                include(pathSpecPredicate._pathSpec);
+            else
+                _predicates.include(p);
+        }
+    }
+
+    /**
+     * {@link IncludeExclude#exclude(Object) Exclude} arbitrary {@link Predicate}s in the conditions.
+     * @param predicates {@link Predicate}s that are tested against the {@link Request}.
+     * This method is optimized so that a passed {@link MethodPredicate} or {@link PathSpecPredicate} is
+     * converted to a more efficient {@link #excludeMethod(String...)} or {@link #exclude(PathSpec...)} respectively.
+     */
+    @SafeVarargs
+    public final void exclude(Predicate<Request>... predicates)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (Predicate<Request> p : predicates)
+        {
+            if (p instanceof MethodPredicate methodPredicate)
+                excludeMethod(methodPredicate._method);
+            else if (p instanceof PathSpecPredicate pathSpecPredicate)
+                exclude(pathSpecPredicate._pathSpec);
+            else
+                _predicates.exclude(p);
+        }
+    }
+
+    private boolean testMethods(Request request)
+    {
+        return _methods.test(request.getMethod());
+    }
+
+    private boolean testPaths(Request request)
+    {
+        return _paths.test(Request.getPathInContext(request));
+    }
+
+    private boolean testPredicates(Request request)
+    {
+        return _predicates.test(request);
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        switch (_conditionNotMetAction)
+        {
+            case SKIP_THIS ->
+            {
+                if (getHandler() == null)
+                    throw new IllegalStateException("No next Handler");
+            }
+            case SKIP_NEXT ->
+            {
+                if (!(getHandler() instanceof Handler.Singleton nextWrapper) || nextWrapper.getHandler() == null)
+                   throw new IllegalStateException("No next Handler.Wrapper next");
+            }
+        }
+
+        super.doStart();
+    }
+
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        // This is the code we would have in handle if MethodHandlers were not used
+        if (testMethods(request) && testPaths(request) && testPredicates(request))
```

Comment:
Suggestion.

If you want to minimize the 3 conditions, even if they are not used, make this a single `Predicate` to test.

Build that single Predicate in `doStart()`.

Make a `MethodsPredicate` (plural) a single `Predicate<Request>` implementation.
Make a `PathsPredicate` (plural) a single `Predicate<Request>` implementation.

Shouldn't be too hard, as `IncludeExcludeSet` is a `Predicate<P>` already.
Don't include the `MethodsPredicate` or `PathsPredicate` in that built `Predicate` if they are unconfigured.

In the example of a totally unconfigured `ConditionalHandler`, you wind up with basically a `(request) -> return true;` Predicate.

Reply:
Example:

``` java
// Declared as a field
public Predicate<Request> handlePredicate = null;
```

``` java
// Code in doStart()
handlePredicate = null;

if (!_methods.isEmpty())
    handlePredicate = (request) -> _methods.test(request.getMethod());

if (!_paths.isEmpty())
{
    Predicate<Request> pathPredicate = (request) -> _paths.test(Request.getPathInContext(request));
    if (handlePredicate != null)
        handlePredicate = handlePredicate.and(pathPredicate);
    else
        handlePredicate = pathPredicate;
}

if (!_predicates.isEmpty())
{
    if (handlePredicate != null)
        handlePredicate = handlePredicate.and(_predicates);
    else
        handlePredicate = _predicates;
}

if (handlePredicate == null)
    handlePredicate = (request) -> true;
```

``` java
// Code in handle()
if (handlePredicate.test(request))
    return doHandle(request, response, callback);        
```

Reply:
Good idea

Reply:
@joakime done, although a bit differently